mlir/lib/Quantum/Utils/RemoveQuantum.cpp

```diff
@@ -37,19 +38,20 @@ void removeQuantumMeasurements(func::FuncOp &function)
     // But the question then becomes, can we have arbitrary control flow
     // inside after measurements?
     //
-    // This will remove the operation in opsToDelete as long as any other uses.
+    // This will remove the operation in opsToDelete as long as it doesn't have any other uses.
     while (!opsToDelete.empty()) {
         Operation *currentOp = opsToDelete.front();
         opsToDelete.pop_front();
-        currentOp->dropAllReferences();
+
+        rewriter.modifyOpInPlace(currentOp, [&] { currentOp->dropAllReferences(); });
         for (Operation *user : currentOp->getUsers()) {
             if (!visited.contains(user)) {
                 visited.insert(user);
                 opsToDelete.push_back(user);
             }
         }
-        if (currentOp->use_empty()) {
-            currentOp->erase();
+        if (currentOp && currentOp->use_empty()) {
+            rewriter.eraseOp(currentOp);
```

Comment:
Nice catch, that was our first memory issue I suppose because this function modified the IR without going through the rewriter 😏